<a href="https://colab.research.google.com/github/vivrelavie/Direcho-Trabaho-DS25B/blob/main/PythonFundamentals_Workbook_I.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Fundamentals
---

- Attendance
- Activity/Mini Projects
- Capstone

---
## Common Python Symbols

- `=` *assignment*
- `==` *equality test*
- `!=` *not equal*

#### Comparisons
- `<`, `<=`, `>`, `>=`

#### Arithmetic
- `+`, `-`, `*`, `/` *add, subtract, multiply, true division (/ keeps float/with decimal)*
- `//` *floor division (drops fractional part)*
- `%` *modulo (remainder)*
- `**` *exponent*
- `+=`, `-=`, `*=`, `/=`, `//=`, `%=` *augmented assignment (ex: `x+=1` == `x = x + 1`)*

#### Logic and Membership
- `and`, `or`, `not`
- `in`, `not in`

## Variables
- names that point to values

In [166]:
# Variables
city = 'Cebu'
year = 2025

## Data Types

In [167]:
# Data Types
integer = 7  ## int
float_num = 3.14 ## float
message = 'hi!' ## str
is_sick = False ## bool

In [168]:
# Checking Data type
type(3.0) ## This is a float -- because of the decimal

float

In [169]:
# Type Conversion
int('42') # str to int
str(3.17) # float to str
float(5) # float to int

5.0

## Basic Python Syntax

In [170]:
city = 'Metro Manila'
print(city)

Metro Manila


In [171]:
if city == 'metro manila':
    print('The same!')
else:
    print('NOT the same!')

NOT the same!


### Printing/Formatting in Python

In [172]:
birthyear = 1997
age = 28

# fstring
print('fstring style')
print(f'I was born in the year {birthyear}')
print(f'I will be {age + 5} in 2030')
print(f'I was born in the year {birthyear}. I will be {age+5} in 2030')
print()
# str.format
print('.format() style')
print('I was born in the year {}'.format(birthyear))
print('I will be {} in 2030'.format(age))
print('I was born in the year {} and I will be {} in 2030'.format(birthyear, age))

fstring style
I was born in the year 1997
I will be 33 in 2030
I was born in the year 1997. I will be 33 in 2030

.format() style
I was born in the year 1997
I will be 28 in 2030
I was born in the year 1997 and I will be 28 in 2030


## Function

In [173]:
# Simple function - converting dollars to peso (USD1 = PHP58.30)
def dollars_to_peso(dollars): # use def keyword + name of function + (parameter)
    return dollars * 58.30

## Comparison and Logical Operations

In [174]:
genz = 28
is_senior = genz >= 60
with_discount = True and is_senior

## Conditionals (IF/ELIF/ELSE)

In [175]:
score = 87
if score >= 95:
    grade = 'A'
elif score >= 80:
    grade = 'B'

## For Loops
- range: creates an immutable (can't be changed) sequence of integers

In [176]:
# with range
## range(n1,n2): Will provide numbers from n1 to n2-1
total = 0
for i in range(1,4): ## 1,2,3
    total += i ## total = total + i

In [177]:
# over a strinng
count = 0
for letter in 'banana':
    if letter == 'a':
        count +=1

In [178]:
# iterate elements on a list (data structure)
list_of_cats = ['1:tuxedo', '2:ginger', '3:tilapia', '4:calico']
for cat in list_of_cats:
    print(cat)

1:tuxedo
2:ginger
3:tilapia
4:calico


------
## Practice Exercises

#### Variable and Types
- Create variables: `first_name`, `age`, `height_m`, `is_happy`. Then print a sentence and provide the type of each.

In [179]:
first_name = 'Stefanie'
age = 20
height_m = 1.47
is_happy = True

print(f"My name is {first_name}. I am {age} years old. My height is {height_m} meters. It's {is_happy} that I am happy.")

My name is Stefanie. I am 20 years old. My height is 1.47 meters. It's True that I am happy.


#### Type Conversion and Printing/Formatting
- Convert '42' (string) to an integer, add 7 then convert the result back to string.

In [180]:
number = int('42')
print(f"Number Type: {type(number)}")
print(f"Number: {number}\n")

number += 7
str_number = str(number)
print(f"Number + 7 = {str_number}")
print(f"Number Type: {type(str_number)}")

Number Type: <class 'int'>
Number: 42

Number + 7 = 49
Number Type: <class 'str'>


#### If conditions
- Write a program to prompt for a score between 0 and 1. If the score is out of range print an error message.
If the score is between 0 and 1, print a grade using the following score-guide equivalent:

|Score   |Grade |
|--------|------|
| >= 0.9 |   A  |
| >= 0.8 |   B  |
| >= 0.7 |   C  |
| >= 0.6 |   D  |
|  < 0.6 |   F  |


In [181]:
"""
Function: 'get_score' will get and validate user input.

Input Validation:
- Check if the input is a number
- Check if the number is between 0 and 1
"""

def get_score():
  score_input = input("Enter your score (Between 0 and 1): ")

  # Check if the input can be converted to a float
  # Try except block added for error handling in case of value mismatch
  try:
    score = float(score_input)
  except ValueError:
    print("Invalid Input. Please enter only numeric values between 0 and 1.")
    return None

  if score < 0 or score > 1:
    print("Input out of range. Please enter only values between 0 and 1.")
    return None
  else:
    return score

In [182]:
score = get_score()

# Check if score is not None before proceeding
if score is not None:
    if score >= 0.9:
        grade = 'A'
    elif score >= 0.8:
        grade = 'B'
    elif score >= 0.7:
        grade = 'C'
    elif score >= 0.6:
        grade = 'D'
    else:
        grade = 'F'
    print(f"Your grade is: {grade}")

Enter your score (Between 0 and 1): 1
Your grade is: A


#### For Loop + Logic
- Sum all numbers from 1 to 20 that are multiples of 3 or 5. Print the sum.

In [183]:
# Given and sum equated to 0 to reset value every run
total = 0
n = 20

for number in range(1,n+1):
  if number % 3 == 0 or number % 5 == 0:
    total += number
    print(f"Number: {number}")

print(f"Final Sum: {total}")

Number: 3
Number: 5
Number: 6
Number: 9
Number: 10
Number: 12
Number: 15
Number: 18
Number: 20
Final Sum: 98


-----
## More Exercises

#### 1. Sales Performance Check
- Given sales_day1 = 150, sales_day2 = 175, sales_day3 = 160, write a function `check_performance()` that:
    - Compares each day’s sales to the previous day.
    - Prints "Better" if the sales are higher, "Worse" if lower, and "Same" if the same.

In [184]:
def check_performance(sales):

  for i in range(1, len(sales)):
    # For readability: sales[i] is the current day sales & sales[i-1] is the previos day sales
    current_day_sales = sales[i]
    previous_day_sales = sales[i-1]

    if current_day_sales > previous_day_sales:
      print(f"Day {i+1} sales: Better")
    elif current_day_sales < previous_day_sales:
      print(f"Day {i+1} sales: Worse")
    else:
      print(f"Day {i+1} sales: Same")

In [185]:
# Given sales values
sales_day1 = 150
sales_day2 = 175
sales_day3 = 160

# sales_day4 = 160

sales_data = [sales_day1, sales_day2, sales_day3]

check_performance(sales_data)

Day 2 sales: Better
Day 3 sales: Worse


#### 2. Simple Discount Calculator
- You are given an item price price = 200 and a discount rate (*discount_rate = 0.15*) (15%). <br>
Write a function `calculate_discount(price, discount_rate)` to:
    - Compute the discount amount (price * discount_rate)
    - Subtract it from the price to find the final price.

In [186]:
def calculate_discount(price, discount_rate):
  discount_amount = price * discount_rate
  final_price = price - discount_amount
  return discount_amount, final_price

In [187]:
# Given values of price and discount rate
price = 200
discount_rate = 0.15

discount_amount, final_price = calculate_discount(price, discount_rate)

print(f"Discount Amount: {discount_amount}")
print(f"Final Price: {final_price}")

Discount Amount: 30.0
Final Price: 170.0


#### 3. Customer Satisfaction Survey
A customer satisfaction score is calculated by averaging responses from 3 surveys. The scores are: <br>
|survery|score|
|-------|------|
|Survey1|4|
|Survey2|3.5|
|Survery3|5|
      
<br>

- Write a function `average_score()` to:
    - Compute the average score.
    - Print whether the score is "Good" (average ≥ 4), "Neutral" (3 ≤ average < 4), or "Bad" (average < 3)

In [192]:
import numpy as np
def average_score(scores):
  # scores_total = sum(scores)
  average = np.sum([float(i) for i in scores]) / len(scores)

  # average = scores_total / len(scores)

  if average >= 4:
    print("Good")
  elif average >= 3:
    print("Neutral")
  else:
    print("Bad")

In [193]:
# Given scores for 3 surveys
Survey1 = 4
Survey2 = 3.5
Survey3 = 5

survey_scores = [Survey1, Survey2, Survey3]

average_score(survey_scores)

Good


#### 4. Revenue vs Target Comparison
The target revenue for the week is target_revenue = 10000. The actual revenue for the week is:
|Revenue Day|Amount|
|-------|------|
|revenue_day1|2000|
|revenue_day2|3000|
|revenue_day3|5000|

- Write a function `check_revenue(target, actual)` to:
    - Compute the total actual revenue.
    - Compare it to the target and print:
        - "On Track" if actual revenue is at least 90% of the target.
        - "Behind" if actual revenue is less than 90% of the target.

In [194]:
def check_revenue(target, actual):
  total_actual = sum(actual)

  if total_actual >= target * 0.9:
    print("On Track")
  else:
    print("Behind")

In [195]:
# Given values for revenue
target_revenue = 10000
revenue_day1 = 2000
revenue_day2 = 3000
revenue_day3 = 5000

actual_revenue = [revenue_day1, revenue_day2, revenue_day3]

check_revenue(target_revenue, actual_revenue)

TypeError: 'int' object is not callable

#### 5. Product Inventory Check
For a product, you are given:
|Tagging|Count|
|-----|------|
|initial_stock|50|
|sold_today|15|
|received_shipment|20|

- Write a function `calculate_remaining_stock()` to:
    - Calculate the remaining stock after today’s sales and the new shipment received.
    - Print "Restocking Needed" if the remaining stock is below 10.
    - Print "Stock is Good" if the remaining stock is at least 10.

#### 6. Daily Expense Summary
You are tracking daily expenses. The daily expenses for the week are:
|Day|Amount Spent (Php)|
|--|--|
|1|250|
|2|185|
|3|450|
|4|170|
|5|348|
|6|70|
|7|99|

- Write a function `average_expenses()` to:
    - Calculate the average daily expense for the week.
    - Print if the expenses are "Within Budget" (average ≤ 250) or "Over Budget" (average >250).

*(You can try to do this with your own daily expenses and the amount that you think should only (your limit) be your daily expense)*

#### 7. Product Profit Calculator
You are selling a product with:
|   |   |
|---|---|
|cost_price|50|
|selling_price|75|

- Write a `function profit_margin()` to:
    - Calculate the profit (selling_price - cost_price)
    - Print "Good Profit" if the profit is at least 20% of the selling price, else print "Low Profit".

*End of Code*